adapters/axis/axis.go

```diff
@@ -0,0 +1,176 @@
+package axis
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint *template.Template
+}
+
+type reqBodyExt struct {
+	AxisBidderExt openrtb_ext.ImpExtAxis `json:"bidder"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpoint: template,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	adapterRequests := make([]*adapters.RequestData, 0, len(request.Imp))
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb2.Imp{imp}
+
+		axisExt, err := a.getImpExt(&imp)
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		url, err := a.buildEndpointURL(axisExt)
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		extJson, err := json.Marshal(reqBodyExt{
+			AxisBidderExt: openrtb_ext.ImpExtAxis{
+				Integration: axisExt.Integration,
+				Token:       axisExt.Token,
+			},
+		})
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		reqCopy.Imp[0].Ext = extJson
+
+		adapterReq, err := a.makeRequest(&reqCopy, url)
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		if adapterReq != nil {
+			adapterRequests = append(adapterRequests, adapterReq)
+		}
+	}
+	return adapterRequests, nil
+}
+
+func (a *adapter) getImpExt(imp *openrtb2.Imp) (*openrtb_ext.ImpExtAxis, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Bidder extension not provided or can't be unmarshalled",
+		}
+	}
+
+	var axisExt openrtb_ext.ImpExtAxis
+	if err := json.Unmarshal(bidderExt.Bidder, &axisExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+
+	return &axisExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ImpExtAxis) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{
+		AccountID: params.Integration,
+		SourceId:  params.Token,
+	}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest, url string) (*adapters.RequestData, error) {
```

Comment:
This function need not be defined on adapter. We can simply define it as:-
```
func makeRequest(request *openrtb2.BidRequest, url string) (*adapters.RequestData, error) {
//your code
}
```

Reply:
Also would like to request if we could change the name of the function to `buildRequest`? This will help in avoiding confusion with the actual `MakeRequest` function that needs to be defined on adapter.